## **AB Testing Meli notebook**

### importando librerias

Se importan las librerias a utilizar durate el proceso de exploración de datos, dentro de las librerias importamos **utilities**, libreria construida específicamente para propósitos de nuestro proyecto.

In [25]:
# Preparamos el entorno
import pandas as pd
import numpy as np
import seaborn as sns; sns.set()
import matplotlib.pyplot as plt
import sys
import os

module_path = os.path.abspath(os.path.join('/Users/juanmanuelpaiba/Documents/Juan_Paiba/AB_Testing_Meli/', 'python_scripts'))
if module_path not in sys.path:
    sys.path.append(module_path)
import utilities # type: ignore

pd.set_option('display.max_rows', None)
pd.set_option('display.max_columns', None)

In [26]:
# Cargamos los datos y creamos el dataframe a utilizar
df = pd.read_csv('/Users/juanmanuelpaiba/Documents/Juan_Paiba/AB_Testing_Meli/data/Inputs/Experiments DataSet For Excercise-small.csv', sep=",")
df.head()

event_name      item_id                     timestamp site  \
0     SEARCH          NaN  2021-08-02T23:55:38.966-0400  MLA   
1    PRODUCT  882352139.0  2021-08-02T23:55:51.673-0400  MLA   
2    PRODUCT  655266729.0  2021-08-02T23:56:16.083-0400  MLA   
3    PRODUCT  761520929.0  2021-08-02T23:56:29.989-0400  MLA   
4    PRODUCT  757586409.0  2021-08-02T23:56:47.887-0400  MLA   

                                         experiments  user_id  
0  {searchbackend/recommended-products=6157, mcli...  3204901  
1  {qadb/sa-on-vip=6695, vip/showV2V3BoxMessages=...  3204901  
2  {qadb/sa-on-vip=6695, vip/showV2V3BoxMessages=...  3204901  
3  {search/remove-ecn-tag=4954, qadb/sa-on-vip=66...  3204901  
4  {search/remove-ecn-tag=4954, qadb/sa-on-vip=66...  3204901

In [42]:
print("Cantidad de registros :" + str(df.shape))
print("Cantidad de usuarios únicos :" + str(df.user_id.nunique()))
print("")
print("Registros por fecha: ")
print(df.date.value_counts().sort_index())
print("")
print("Usuarios unicos por fecha: " + str(df.groupby('date')['user_id'].nunique()))
print("")
print("Secciones del sitio navegadas: ")
print(df.event_name.value_counts())
print(df.site.value_counts())

Cantidad de registros :(141553, 8)
Cantidad de usuarios únicos :7817

Registros por fecha: 
date
2021-08-01     31094
2021-08-02    110459
Name: count, dtype: int64

Usuarios unicos por fecha: date
2021-08-01    2138
2021-08-02    6735
Name: user_id, dtype: int64

Secciones del sitio navegadas: 
event_name
SEARCH        76750
PRODUCT       58544
CHECKOUT_1     2115
CHECKOUT_2     1786
CHECKOUT_3     1270
BUY            1088
Name: count, dtype: int64
site
MLA    141553
Name: count, dtype: int64


Control de registros de las variables disponibles, tenemos un aproximado de 150k registros o navegaciones, para el transcurso de dos días 202108 (01-02), para 7,817 clientes. Eventualmente se observa que a medida se avanza en el proceso de compra las porciones de navegación disminuyes significativamente, donde 76750 fueron eventos relacionados con la búsqueda y solamente 1,088 de compra. 

In [40]:
# variable date, tipo fecha
df["datetime"] = pd.to_datetime(df["timestamp"], format="%Y-%m-%dT%H:%M:%S.%f%z")
df["date"] = pd.to_datetime(df["datetime"]).dt.date
print(df.dtypes)


event_name                        object
item_id                          float64
timestamp                         object
site                              object
experiments                       object
user_id                            int64
datetime       datetime64[ns, UTC-04:00]
date                              object
dtype: object


In [44]:
df[df.event_name == "SEARCH"].head()

event_name  item_id                     timestamp site  \
0      SEARCH      NaN  2021-08-02T23:55:38.966-0400  MLA   
6      SEARCH      NaN  2021-08-02T23:46:03.785-0400  MLA   
7      SEARCH      NaN  2021-08-02T23:46:39.972-0400  MLA   
8      SEARCH      NaN  2021-08-02T23:47:13.438-0400  MLA   
10     SEARCH      NaN  2021-08-02T23:39:56.973-0400  MLA   

                                          experiments  user_id  \
0   {searchbackend/recommended-products=6157, mcli...  3204901   
6   {searchbackend/recommended-products=6157, mcli...  4484451   
7   {searchbackend/recommended-products=6157, mcli...  4484451   
8   {searchbackend/recommended-products=6157, mcli...  4484451   
10  {searchbackend/recommended-products=6157, sear...   402295   

                           datetime        date  
0  2021-08-02 23:55:38.966000-04:00  2021-08-02  
6  2021-08-02 23:46:03.785000-04:00  2021-08-02  
7  2021-08-02 23:46:39.972000-04:00  2021-08-02  
8  2021-08-02 23:47:13.438000-04:00  2021-08-02  
10 2021-08-02 23:39:56.973000-04:00  2021-08-02

In [43]:
df[df.event_name == "BUY"].head()

event_name      item_id                     timestamp site  \
204        BUY  740440936.0  2021-08-02T23:53:46.363-0400  MLA   
295        BUY  680029039.0  2021-08-02T23:38:37.024-0400  MLA   
305        BUY  988145209.0  2021-08-02T23:41:31.971-0400  MLA   
447        BUY  885565729.0  2021-08-02T23:11:25.388-0400  MLA   
562        BUY  934373429.0  2021-08-02T20:01:39.802-0400  MLA   

                                           experiments  user_id  \
204  {buyingflow/user-track=6796, buyingflow/addres...   426144   
295  {buyingflow/user-track=6796, buyingflow/addres...  6498883   
305  {buyingflow/user-track=6796, buyingflow/addres...  6498883   
447  {buyingflow/user-track=6796, buyingflow/addres...  2866832   
562  {buyingflow/user-track=6796, buyingflow/addres...   411939   

                            datetime        date  
204 2021-08-02 23:53:46.363000-04:00  2021-08-02  
295 2021-08-02 23:38:37.024000-04:00  2021-08-02  
305 2021-08-02 23:41:31.971000-04:00  2021-08-02  
447 2021-08-02 23:11:25.388000-04:00  2021-08-02  
562 2021-08-02 20:01:39.802000-04:00  2021-08-02

In [45]:
df[df.user_id == 3204901].sort_values(by='datetime').head(5)

event_name      item_id                     timestamp site  \
63670     SEARCH          NaN  2021-08-02T12:53:11.057-0400  MLA   
63671     SEARCH          NaN  2021-08-02T12:53:20.260-0400  MLA   
28364     SEARCH          NaN  2021-08-02T16:16:05.995-0400  MLA   
28365    PRODUCT  312247919.0  2021-08-02T16:16:32.788-0400  MLA   
28366    PRODUCT  343345729.0  2021-08-02T16:16:45.106-0400  MLA   

                                             experiments  user_id  \
63670  {search/remove-ecn-tag=4954, searchbackend/rec...  3204901   
63671  {searchbackend/recommended-products=6157, sear...  3204901   
28364  {searchbackend/recommended-products=6157, sear...  3204901   
28365  {search/remove-ecn-tag=4954, vip/showV2V3BoxMe...  3204901   
28366  {search/remove-ecn-tag=4954, vip/showV2V3BoxMe...  3204901   

                              datetime        date  
63670 2021-08-02 12:53:11.057000-04:00  2021-08-02  
63671 2021-08-02 12:53:20.260000-04:00  2021-08-02  
28364 2021-08-02 16:16:05.995000-04:00  2021-08-02  
28365 2021-08-02 16:16:32.788000-04:00  2021-08-02  
28366 2021-08-02 16:16:45.106000-04:00  2021-08-02

Imprimimos algunos de los registros de compra y búsqueda de nuestros clientes, y las primeras sesiones de navegación para un cliente cualquiera, observando que para los **event_name SEARCH o ventos de búsqueda no es posible identificar un item_id**, suceso de particular interés, puesto que para realizar seguimiento completo a la cadena de compra vamos a utilizar el item_id o artículo de interés.

Revisaremos en detalle algunos de los procesos de compra para nuestro clientes: 

In [57]:
df[df.user_id == 6498883].sort_values(by='datetime').loc[299:305]
#df[df.user_id == 426144].sort_values(by='datetime').loc[200:204]
#df[df.user_id == 466].sort_values(by='datetime')

event_name      item_id                     timestamp site  \
299      SEARCH          NaN  2021-08-02T23:39:55.540-0400  MLA   
300     PRODUCT  988145209.0  2021-08-02T23:39:56.097-0400  MLA   
301  CHECKOUT_1  988145209.0  2021-08-02T23:40:30.328-0400  MLA   
302  CHECKOUT_2  988145209.0  2021-08-02T23:40:40.135-0400  MLA   
303  CHECKOUT_2  988145209.0  2021-08-02T23:40:51.520-0400  MLA   
304  CHECKOUT_3  988145209.0  2021-08-02T23:41:18.970-0400  MLA   
305         BUY  988145209.0  2021-08-02T23:41:31.971-0400  MLA   

                                           experiments  user_id  \
299  {searchbackend/recommended-products=6157, sear...  6498883   
300  {search/remove-ecn-tag=4954, qadb/sa-on-vip=66...  6498883   
301  {buyingflow/user-track=6796, buyingflow/escWeb...  6498883   
302  {buyingflow/user-track=6796, buyingflow/addres...  6498883   
303                       {buyingflow/user-track=6796}  6498883   
304  {buyingflow/user-track=6796, buyingflow/addres...  6498883   
305  {buyingflow/user-track=6796, buyingflow/addres...  6498883   

                            datetime        date  
299 2021-08-02 23:39:55.540000-04:00  2021-08-02  
300 2021-08-02 23:39:56.097000-04:00  2021-08-02  
301 2021-08-02 23:40:30.328000-04:00  2021-08-02  
302 2021-08-02 23:40:40.135000-04:00  2021-08-02  
303 2021-08-02 23:40:51.520000-04:00  2021-08-02  
304 2021-08-02 23:41:18.970000-04:00  2021-08-02  
305 2021-08-02 23:41:31.971000-04:00  2021-08-02

In [56]:
print(df[df.user_id == 6498883].experiments[299].strip('{}').split(', '))
print(df[df.user_id == 6498883].experiments[300].strip('{}').split(', '))
print(df[df.user_id == 6498883].experiments[301].strip('{}').split(', '))
print(df[df.user_id == 6498883].experiments[302].strip('{}').split(', '))
print(df[df.user_id == 6498883].experiments[303].strip('{}').split(', '))
print(df[df.user_id == 6498883].experiments[304].strip('{}').split(', '))

['searchbackend/recommended-products=6157', 'search/ungroup-products=3273', 'searchbackend/cbt-antiboost=6333', 'search/back-filters=5059', 'filters/sort-by-ranking=7057', 'search/best-seller-aa-testing-fail-fast-edition=4514', 'frontend/assetsCdnDomainMLU=DEFAULT', 'searchbackend/item-reputation=3824', 'search/remove-ecn-tag=4954', 'searchbackend/seller-reputation-change=4553', 'cookiesConsentBanner=DEFAULT', 'frontend/assetsCdnDomainMLA=DEFAULT', 'search/best-seller-fail-fast-edition-MLA=4916']
['search/remove-ecn-tag=4954', 'qadb/sa-on-vip=6696', 'vip/showV2V3BoxMessages=6430', 'search/tendency-landing-enabled-MLA=6980', 'pdp/viewItemPageMigrationDesktopReviewsNoTabs=4856', 'cookiesConsentBanner=DEFAULT', 'vip/shippingCalculatorMigrationModalExperiment=6551', 'vip/carousel-v2p-above-the-fold=6787', 'pdp/viewItemPageMigrationReturns=5208', 'vip/seller-with-tooltip=4692']
['buyingflow/user-track=6796', 'buyingflow/escWebMLA=2874']
['buyingflow/user-track=6796', 'buyingflow/address_hub

In [58]:
## eventos detectados durante la sesión de navegación Search
df[df.user_id == 6498883].experiments[299].strip('{}').split(', ')

['searchbackend/recommended-products=6157',
 'search/ungroup-products=3273',
 'searchbackend/cbt-antiboost=6333',
 'search/back-filters=5059',
 'filters/sort-by-ranking=7057',
 'search/best-seller-aa-testing-fail-fast-edition=4514',
 'frontend/assetsCdnDomainMLU=DEFAULT',
 'searchbackend/item-reputation=3824',
 'search/remove-ecn-tag=4954',
 'searchbackend/seller-reputation-change=4553',
 'cookiesConsentBanner=DEFAULT',
 'frontend/assetsCdnDomainMLA=DEFAULT',
 'search/best-seller-fail-fast-edition-MLA=4916']

In [59]:
## eventos detectados durante la sesión de navegación  Product
df[df.user_id == 6498883].experiments[300].strip('{}').split(', ')

['search/remove-ecn-tag=4954',
 'qadb/sa-on-vip=6696',
 'vip/showV2V3BoxMessages=6430',
 'search/tendency-landing-enabled-MLA=6980',
 'pdp/viewItemPageMigrationDesktopReviewsNoTabs=4856',
 'cookiesConsentBanner=DEFAULT',
 'vip/shippingCalculatorMigrationModalExperiment=6551',
 'vip/carousel-v2p-above-the-fold=6787',
 'pdp/viewItemPageMigrationReturns=5208',
 'vip/seller-with-tooltip=4692']

Validando el seguimiento de la cadena de compra de nuestro cliente **6498883** y su producto adquirido vía mercado libre **988145209**, confirmamos que :
 * eventos de **search** no tienen un ítem relacionado
 * dada la cantidad de eventos realizados durante la búsqueda, y el interés particular por determinar cual de las variantes aplicadas tuvo mayor éxito de compra, es necesario mapear o marcar cada uno de los mismos con una etiqueta positiva (1). 

### **Marcación de experimentos y variantes exitosas en cadena de compra**

Tomando en consideración todo lo mencionado anteriormente, y dando respuesta a la primera solicitud de nuestro ejercicio, buscaremos asignar de la manera más precisa, todos los eventos (y sus variantes correspondientes) involucrados en la cadena de cada una de las compras, tomando en consideración desde la navegación de búsqueda hasta la de compra.

In [20]:
# Find all item_id and user_id pairs that end with a 'BUY' event
buy_pairs = df[df['event_name'] == 'BUY'][['item_id', 'user_id']].drop_duplicates()

# Define a set for quick lookup
buy_pairs_set = set(tuple(x) for x in buy_pairs.values)

# Function to determine if the row is in the purchase funnel
def in_purchase_funnel(row):
    if pd.notna(row['item_id']):  # Only consider rows with a valid item_id
        return 1 if (row['item_id'], row['user_id']) in buy_pairs_set else 0
    return 0

# Apply the function to create the purchase_funnel_flag column
df['purchase_funnel_flag'] = df.apply(in_purchase_funnel, axis=1)

Creamos entonces una variable **purchase_funnel_flag**, que nos permita asignar 1 a todos aquellos experimentos durante la cadena de compra por los cuales paso nuestro cliente, llegando así a una **compra exitosa**

In [21]:
# Sort the DataFrame by user_id and datetime
df_sorted = df.sort_values(by=['user_id', 'datetime']).reset_index(drop=True)

# Create a mask for rows where purchase_funnel_flag is 1
mask = df_sorted['purchase_funnel_flag'] == 1

# Create a shifted version of the event_name column
df_sorted['prev_event'] = df_sorted['event_name'].shift(1)
df_sorted['prev_user'] = df_sorted['user_id'].shift(1).fillna(0).astype(int)
df_sorted['prev_date'] = df_sorted['date'].shift(1)

# Create a mask for rows where the previous event is SEARCH and user_id matches
search_mask = (df_sorted['prev_event'] == 'SEARCH') & (df_sorted['user_id'] == df_sorted['prev_user']) & (df_sorted['date'] == df_sorted['prev_date']) & (df_sorted['purchase_funnel_flag'] == 1)
search_mask_1 = search_mask.shift(-1)

# Update purchase_funnel_flag where mask is True and search_mask is True
df_sorted.loc[mask | search_mask_1, 'purchase_funnel_flag'] = 1

# Drop the 'prev_event' column
df_sorted.drop(columns=['prev_event','prev_user','prev_date'], inplace=True)

In [22]:
#df_sorted[df_sorted.user_id == 466].sort_values(by='datetime')
df_sorted[df_sorted.user_id == 6498883].sort_values(by='datetime')

event_name      item_id                     timestamp site  \
117111      SEARCH          NaN  2021-08-01T20:47:19.528-0400  MLA   
117112      SEARCH          NaN  2021-08-01T20:54:02.302-0400  MLA   
117113  CHECKOUT_1  680029039.0  2021-08-02T23:37:16.248-0400  MLA   
117114  CHECKOUT_2  680029039.0  2021-08-02T23:37:28.379-0400  MLA   
117115  CHECKOUT_3  680029039.0  2021-08-02T23:38:14.079-0400  MLA   
117116         BUY  680029039.0  2021-08-02T23:38:37.024-0400  MLA   
117117  CHECKOUT_2  680029039.0  2021-08-02T23:39:08.456-0400  MLA   
117118  CHECKOUT_2  680029039.0  2021-08-02T23:39:38.513-0400  MLA   
117119  CHECKOUT_2  680029039.0  2021-08-02T23:39:42.170-0400  MLA   
117120      SEARCH          NaN  2021-08-02T23:39:55.540-0400  MLA   
117121     PRODUCT  988145209.0  2021-08-02T23:39:56.097-0400  MLA   
117122  CHECKOUT_1  988145209.0  2021-08-02T23:40:30.328-0400  MLA   
117123  CHECKOUT_2  988145209.0  2021-08-02T23:40:40.135-0400  MLA   
117124  CHECKOUT_2  988145209.0  2021-08-02T23:40:51.520-0400  MLA   
117125  CHECKOUT_3  988145209.0  2021-08-02T23:41:18.970-0400  MLA   
117126         BUY  988145209.0  2021-08-02T23:41:31.971-0400  MLA   
117127     PRODUCT  988145209.0  2021-08-02T23:41:49.952-0400  MLA   

                                              experiments  user_id  \
117111  {searchbackend/recommended-products=6157, sear...  6498883   
117112  {searchbackend/recommended-products=6157, sear...  6498883   
117113  {buyingflow/user-track=6796, buyingflow/escWeb...  6498883   
117114  {buyingflow/user-track=6796, buyingflow/addres...  6498883   
117115  {buyingflow/user-track=6796, buyingflow/addres...  6498883   
117116  {buyingflow/user-track=6796, buyingflow/addres...  6498883   
117117                       {buyingflow/user-track=6796}  6498883   
117118                       {buyingflow/user-track=6796}  6498883   
117119                       {buyingflow/user-track=6796}  6498883   
117120  {searchbackend/recommended-products=6157, sear...  6498883   
117121  {search/remove-ecn-tag=4954, qadb/sa-on-vip=66...  6498883   
117122  {buyingflow/user-track=6796, buyingflow/escWeb...  6498883   
117123  {buyingflow/user-track=6796, buyingflow/addres...  6498883   
117124                       {buyingflow/user-track=6796}  6498883   
117125  {buyingflow/user-track=6796, buyingflow/addres...  6498883   
117126  {buyingflow/user-track=6796, buyingflow/addres...  6498883   
117127  {search/remove-ecn-tag=4954, qadb/sa-on-vip=66...  6498883   

                               datetime        date  purchase_funnel_flag  
117111 2021-08-01 20:47:19.528000-04:00  2021-08-01                     0  
117112 2021-08-01 20:54:02.302000-04:00  2021-08-01                     0  
117113 2021-08-02 23:37:16.248000-04:00  2021-08-02                     1  
117114 2021-08-02 23:37:28.379000-04:00  2021-08-02                     1  
117115 2021-08-02 23:38:14.079000-04:00  2021-08-02                     1  
117116 2021-08-02 23:38:37.024000-04:00  2021-08-02                     1  
117117 2021-08-02 23:39:08.456000-04:00  2021-08-02                     1  
117118 2021-08-02 23:39:38.513000-04:00  2021-08-02                     1  
117119 2021-08-02 23:39:42.170000-04:00  2021-08-02                     1  
117120 2021-08-02 23:39:55.540000-04:00  2021-08-02                     1  
117121 2021-08-02 23:39:56.097000-04:00  2021-08-02                     1  
117122 2021-08-02 23:40:30.328000-04:00  2021-08-02                     1  
117123 2021-08-02 23:40:40.135000-04:00  2021-08-02                     1  
117124 2021-08-02 23:40:51.520000-04:00  2021-08-02                     1  
117125 2021-08-02 23:41:18.970000-04:00  2021-08-02                     1  
117126 2021-08-02 23:41:31.971000-04:00  2021-08-02                     1  
117127 2021-08-02 23:41:49.952000-04:00  2021-08-02                     1

In [23]:
print(df.purchase_funnel_flag.value_counts())
print(df_sorted.purchase_funnel_flag.value_counts())

purchase_funnel_flag
0    134867
1      6686
Name: count, dtype: int64
purchase_funnel_flag
0    134068
1      7485
Name: count, dtype: int64


Procesamiento de la información, inicialmente convertimos la columna **experiments** a diccionarios, y luego realizamos la apertira de cada uno de los mismos para obener por fila cada uno de los eventos,y limpieza de columnas que ya no utilizaremos para las finalidades del ejerccio

In [24]:
# Apply the function to the 'experiments' column and create a new column with the dictionaries
df_sorted['experiments_dict'] = df_sorted['experiments'].apply(utilities.str_to_dict)

# aplicamos funcion explode_dict para apertura de diccionarios
df_exploded = df_sorted.apply(utilities.explode_dict, axis=1).reset_index(drop=True)
df_exploded = pd.concat(df_exploded.to_list(), ignore_index=True)

# Drop the 'experiments_dict' column as it is no longer needed
df_exploded.drop(columns=['experiments_dict','timestamp','experiments'], inplace=True)

# separamos columna experiment, para obtener su correspondiente descripción y variante
df_exploded[['path', 'experiment']] = df_exploded['experiment_key'].str.split('/', expand=True)
df_exploded.rename(columns={'experiment_value': 'variant'}, inplace=True)

KeyboardInterrupt: 

In [ ]:
#df_exploded[(df_exploded.user_id == 466)&(df_exploded.experiment == 'search-list-algorithms')].sort_values(by='datetime')
df_exploded[(df_exploded.user_id == 466)].sort_values(by='datetime')

event_name      item_id site  user_id                         datetime  \
1766      SEARCH          NaN  MLA      466 2021-08-02 19:44:09.487000-04:00   
1782      SEARCH          NaN  MLA      466 2021-08-02 19:44:09.487000-04:00   
1781      SEARCH          NaN  MLA      466 2021-08-02 19:44:09.487000-04:00   
1780      SEARCH          NaN  MLA      466 2021-08-02 19:44:09.487000-04:00   
1779      SEARCH          NaN  MLA      466 2021-08-02 19:44:09.487000-04:00   
1778      SEARCH          NaN  MLA      466 2021-08-02 19:44:09.487000-04:00   
1777      SEARCH          NaN  MLA      466 2021-08-02 19:44:09.487000-04:00   
1775      SEARCH          NaN  MLA      466 2021-08-02 19:44:09.487000-04:00   
1776      SEARCH          NaN  MLA      466 2021-08-02 19:44:09.487000-04:00   
1773      SEARCH          NaN  MLA      466 2021-08-02 19:44:09.487000-04:00   
1772      SEARCH          NaN  MLA      466 2021-08-02 19:44:09.487000-04:00   
1771      SEARCH          NaN  MLA      466 2021-08-02 19:44:09.487000-04:00   
1770      SEARCH          NaN  MLA      466 2021-08-02 19:44:09.487000-04:00   
1769      SEARCH          NaN  MLA      466 2021-08-02 19:44:09.487000-04:00   
1768      SEARCH          NaN  MLA      466 2021-08-02 19:44:09.487000-04:00   
1767      SEARCH          NaN  MLA      466 2021-08-02 19:44:09.487000-04:00   
1774      SEARCH          NaN  MLA      466 2021-08-02 19:44:09.487000-04:00   
1788      SEARCH          NaN  MLA      466 2021-08-02 19:44:16.693000-04:00   
1791      SEARCH          NaN  MLA      466 2021-08-02 19:44:16.693000-04:00   
1790      SEARCH          NaN  MLA      466 2021-08-02 19:44:16.693000-04:00   
1787      SEARCH          NaN  MLA      466 2021-08-02 19:44:16.693000-04:00   
1789      SEARCH          NaN  MLA      466 2021-08-02 19:44:16.693000-04:00   
1785      SEARCH          NaN  MLA      466 2021-08-02 19:44:16.693000-04:00   
1786      SEARCH          NaN  MLA      466 2021-08-02 19:44:16.693000-04:00   
1784      SEARCH          NaN  MLA      466 2021-08-02 19:44:16.693000-04:00   
1783      SEARCH          NaN  MLA      466 2021-08-02 19:44:16.693000-04:00   
1799      SEARCH          NaN  MLA      466 2021-08-02 19:44:23.557000-04:00   
1798      SEARCH          NaN  MLA      466 2021-08-02 19:44:23.557000-04:00   
1797      SEARCH          NaN  MLA      466 2021-08-02 19:44:23.557000-04:00   
1796      SEARCH          NaN  MLA      466 2021-08-02 19:44:23.557000-04:00   
1800      SEARCH          NaN  MLA      466 2021-08-02 19:44:23.557000-04:00   
1794      SEARCH          NaN  MLA      466 2021-08-02 19:44:23.557000-04:00   
1793      SEARCH          NaN  MLA      466 2021-08-02 19:44:23.557000-04:00   
1792      SEARCH          NaN  MLA      466 2021-08-02 19:44:23.557000-04:00   
1795      SEARCH          NaN  MLA      466 2021-08-02 19:44:23.557000-04:00   
1809      SEARCH          NaN  MLA      466 2021-08-02 19:44:30.785000-04:00   
1808      SEARCH          NaN  MLA      466 2021-08-02 19:44:30.785000-04:00   
1807      SEARCH          NaN  MLA      466 2021-08-02 19:44:30.785000-04:00   
1806      SEARCH          NaN  MLA      466 2021-08-02 19:44:30.785000-04:00   
1805      SEARCH          NaN  MLA      466 2021-08-02 19:44:30.785000-04:00   
1803      SEARCH          NaN  MLA      466 2021-08-02 19:44:30.785000-04:00   
1802      SEARCH          NaN  MLA      466 2021-08-02 19:44:30.785000-04:00   
1801      SEARCH          NaN  MLA      466 2021-08-02 19:44:30.785000-04:00   
1804      SEARCH          NaN  MLA      466 2021-08-02 19:44:30.785000-04:00   
1815      SEARCH          NaN  MLA      466 2021-08-02 19:45:00.518000-04:00   
1818      SEARCH          NaN  MLA      466 2021-08-02 19:45:00.518000-04:00   
1816      SEARCH          NaN  MLA      466 2021-08-02 19:45:00.518000-04:00   
1814      SEARCH          NaN  MLA      466 2021-08-02 19:45:00.518000-04:00   
1817      SEARCH          NaN  MLA      466 2021-08-02 19:45:00.518000-04:00   
1812  

### 1. **Agrupación de la información para realizar el conteo**

In [ ]:
df_exploded['item_id'] = df_exploded['item_id'].fillna(0)

# Agrupamos con variables necesarias para conteo
df_grouped = df_exploded.groupby(['date','experiment','variant','user_id','path']).agg({'purchase_funnel_flag': 'max',  'item_id': lambda x: x[df_exploded['purchase_funnel_flag'] != 0].nunique() }).reset_index()

df_grouped.head()

date           experiment variant  user_id    path  \
0  2021-08-01  HideTransitionModal    6646   336575  mshops   
1  2021-08-01  HideTransitionModal    6646  3677574  mshops   
2  2021-08-01  HideTransitionModal    6647  2234611  mshops   
3  2021-08-01  HideTransitionModal    6647  3023131  mshops   
4  2021-08-01  HideTransitionModal    6647  9116431  mshops   

   purchase_funnel_flag  item_id  
0                     0        0  
1                     0        0  
2                     0        0  
3                     0        0  
4                     0        0

QUe hacer cuando un cliente compra varias veces el mismo artículo ?

In [ ]:
df_exploded[df_exploded.user_id == 6498883].sort_values(by='datetime').head(50)
#df_exploded[df_exploded.user_id == 466].sort_values(by='datetime')


event_name      item_id site  user_id  \
1107071      SEARCH          0.0  MLA  6498883   
1107088      SEARCH          0.0  MLA  6498883   
1107087      SEARCH          0.0  MLA  6498883   
1107086      SEARCH          0.0  MLA  6498883   
1107085      SEARCH          0.0  MLA  6498883   
1107083      SEARCH          0.0  MLA  6498883   
1107082      SEARCH          0.0  MLA  6498883   
1107081      SEARCH          0.0  MLA  6498883   
1107080      SEARCH          0.0  MLA  6498883   
1107084      SEARCH          0.0  MLA  6498883   
1107078      SEARCH          0.0  MLA  6498883   
1107077      SEARCH          0.0  MLA  6498883   
1107076      SEARCH          0.0  MLA  6498883   
1107075      SEARCH          0.0  MLA  6498883   
1107074      SEARCH          0.0  MLA  6498883   
1107073      SEARCH          0.0  MLA  6498883   
1107072      SEARCH          0.0  MLA  6498883   
1107079      SEARCH          0.0  MLA  6498883   
1107100      SEARCH          0.0  MLA  6498883   
1107101      SEARCH          0.0  MLA  6498883   
1107102      SEARCH          0.0  MLA  6498883   
1107106      SEARCH          0.0  MLA  6498883   
1107104      SEARCH          0.0  MLA  6498883   
1107105      SEARCH          0.0  MLA  6498883   
1107099      SEARCH          0.0  MLA  6498883   
1107103      SEARCH          0.0  MLA  6498883   
1107098      SEARCH          0.0  MLA  6498883   
1107092      SEARCH          0.0  MLA  6498883   
1107096      SEARCH          0.0  MLA  6498883   
1107095      SEARCH          0.0  MLA  6498883   
1107094      SEARCH          0.0  MLA  6498883   
1107093      SEARCH          0.0  MLA  6498883   
1107091      SEARCH          0.0  MLA  6498883   
1107090      SEARCH          0.0  MLA  6498883   
1107089      SEARCH          0.0  MLA  6498883   
1107097      SEARCH          0.0  MLA  6498883   
1107107  CHECKOUT_1  680029039.0  MLA  6498883   
1107108  CHECKOUT_1  680029039.0  MLA  6498883   
1107109  CHECKOUT_2  680029039.0  MLA  6498883   
1107110  CHECKOUT_2  680029039.0  MLA  6498883   
1107111  CHECKOUT_3  680029039.0  MLA  6498883   
1107112  CHECKOUT_3  680029039.0  MLA  6498883   
1107113         BUY  680029039.0  MLA  6498883   
1107114         BUY  680029039.0  MLA  6498883   
1107115  CHECKOUT_2  680029039.0  MLA  6498883   
1107116  CHECKOUT_2  680029039.0  MLA  6498883   
1107117  CHECKOUT_2  680029039.0  MLA  6498883   
1107130      SEARCH          0.0  MLA  6498883   
1107129      SEARCH          0.0  MLA  6498883   
1107128      SEARCH          0.0  MLA  6498883   

                                datetime        date  purchase_funnel_flag  \
1107071 2021-08-01 20:47:19.528000-04:00  2021-08-01                     0   
1107088 2021-08-01 20:47:19.528000-04:00  2021-08-01                     0   
1107087 2021-08-01 20:47:19.528000-04:00  2021-08-01                     0   
1107086 2021-08-01 20:47:19.528000-04:00  2021-08-01                     0   
1107085 2021-08-01 20:47:19.528000-04:00  2021-08-01                     0   
1107083 2021-08-01 20:47:19.528000-04:00  2021-08-01                     0   
1107082 2021-08-01 20:47:19.528000-04:00  2021-08-01                     0   
1107081 2021-08-01 20:47:19.528000-04:00  2021-08-01                     0   
1107080 2021-08-01 20:47:19.528000-04:00  2021-08-01                     0   
1107084 2021-08-01 20:47:19.528000-04:00  2021-08-01                     0   
1107078 2021-08-01 20:47:19.528000-04:00  2021-08-01                     0   
1107077 2021-08-01 20:47:19.528000-04:00  2021-08-01                     0   
1107076 2021-08-01 20:47:19.528000-04:00  2021-08-01                     0   
1107075 2021-08-01 20:47:19.528000-04:00  2021-08-01                     0   
1107074 2021-08-01 20:47:19.528000-04:00  2021-08-01                     0   
1107073 2021-08-01 20:47:19.528000-04:00  2021-08-01                     0   
1107072 2021-08-01 20:47:19.528000-04:00  2021-08-01                     0   
1107079 2021-08-01 20:47:19.528000-04:00  2021-08-01  

In [ ]:
#df_grouped[df_grouped.user_id == 466].sort_values(by='experiment')
df_grouped[df_grouped.user_id == 6498883].sort_values(by='date').head(20)

date                                  experiment  variant  \
1162    2021-08-01  ads-adsearch-boost-incremental-desktop-mla     3809   
33297   2021-08-01                            ungroup-products     3273   
31845   2021-08-01                             sort-by-ranking     7057   
28337   2021-08-01                       show-pads-search-list     5146   
22942   2021-08-01                    seller-reputation-change     4553   
21186   2021-08-01                search-pads-none-desktop-mla     3478   
20018   2021-08-01                      search-list-algorithms     5528   
18052   2021-08-01                              remove-ecn-tag     4954   
27144   2021-08-01                            show-pads-global     5176   
14276   2021-08-01                             item-reputation     3824   
12207   2021-08-01                               cbt-antiboost     6333   
9189    2021-08-01           best-seller-fail-fast-edition-MLA     4916   
7268    2021-08-01    best-seller-aa-testing-fail-fast-edition     4514   
6344    2021-08-01                                back-filters     5059   
4580    2021-08-01                          assetsCdnDomainMLU  DEFAULT   
2815    2021-08-01                          assetsCdnDomainMLA  DEFAULT   
15496   2021-08-01                        recommended-products     6157   
111554  2021-08-02                    seller-reputation-change     4553   
116364  2021-08-02                         seller-with-tooltip     4692   
121055  2021-08-02  shippingCalculatorMigrationModalExperiment     6551   

        user_id           path  purchase_funnel_flag  item_id  
1162    6498883         mclics                     0        0  
33297   6498883         search                     0        0  
31845   6498883        filters                     0        0  
28337   6498883         mclics                     0        0  
22942   6498883  searchbackend                     0        0  
21186   6498883         mclics                     0        0  
20018   6498883         mclics                     0        0  
18052   6498883         search                     0        0  
27144   6498883         mclics                     0        0  
14276   6498883  searchbackend                     0        0  
12207   6498883  searchbackend                     0        0  
9189    6498883         search                     0        0  
7268    6498883         search                     0        0  
6344    6498883         search                     0        0  
4580    6498883       frontend                     0        0  
2815    6498883       frontend                     0        0  
15496   6498883  searchbackend                     0        0  
111554  6498883  searchbackend                     1        1  
116364  6498883            vip                     1        1  
121055  6498883            vip                     1        1

In [ ]:
df_grouped_00 = df_grouped.groupby(['date', 'experiment', 'variant']).agg({'user_id': 'count', 'purchase_funnel_flag': 'sum'}).reset_index()
df_grouped_00 = df_grouped_00.rename(columns={"user_id": "participants" , "purchase_funnel_flag": "purchases"})
df_grouped_00['buy_rate'] = df_grouped_00 ['purchases']/df_grouped_00 ['participants']
df_grouped_00['buy_rate_percent'] = df_grouped_00['buy_rate'] .apply(lambda x: '{:.2f}%'.format(x*100))
#df_5 = df_5.sort_values(['date','experiment_name', 'conversion_rate_%', ], ascending = [True, True, False], ignore_index=True)
#df_5


In [ ]:
df_grouped_00.shape

(143, 7)

In [ ]:
df_grouped_00.head(130)

date                                  experiment  variant  \
0    2021-08-01                         HideTransitionModal     6646   
1    2021-08-01                         HideTransitionModal     6647   
2    2021-08-01                         HideTransitionModal     6993   
3    2021-08-01                         HideTransitionModal  DEFAULT   
4    2021-08-01                                 address_hub     3574   
5    2021-08-01  ads-adsearch-boost-incremental-desktop-mla     3809   
6    2021-08-01                          assetsCdnDomainMLA  DEFAULT   
7    2021-08-01                          assetsCdnDomainMLU  DEFAULT   
8    2021-08-01                                back-filters     5059   
9    2021-08-01                                back-filters  DEFAULT   
10   2021-08-01    best-seller-aa-testing-fail-fast-edition     4514   
11   2021-08-01    best-seller-aa-testing-fail-fast-edition     4515   
12   2021-08-01    best-seller-aa-testing-fail-fast-edition  DEFAULT   
13   2021-08-01           best-seller-fail-fast-edition-MLA     4916   
14   2021-08-01           best-seller-fail-fast-edition-MLA  DEFAULT   
15   2021-08-01                 carousel-v2p-above-the-fold     6786   
16   2021-08-01                 carousel-v2p-above-the-fold     6787   
17   2021-08-01                 carousel-v2p-above-the-fold  DEFAULT   
18   2021-08-01                               cbt-antiboost     6333   
19   2021-08-01                             checkOnBehavior      348   
20   2021-08-01                             checkOnBehavior      349   
21   2021-08-01                           classiWordingFree     2826   
22   2021-08-01                         compatibilityWidget     5539   
23   2021-08-01                        cpgShowOnlyAddToCart     6690   
24   2021-08-01                           enable-map-layout     5929   
25   2021-08-01                                   escWebMLA     2874   
26   2021-08-01                             item-reputation     3824   
27   2021-08-01                 official-store-orders-boost     4313   
28   2021-08-01                 official-store-orders-boost     4314   
29   2021-08-01                             official-stores     4926   
30   2021-08-01                        recommended-products     6157   
31   2021-08-01                        recommended-products     6158   
32   2021-08-01                              remove-ecn-tag     4954   
33   2021-08-01                              remove-ecn-tag  DEFAULT   
34   2021-08-01                   results-target-web-motors     7327   
35   2021-08-01                   results-target-web-motors     7328   
36   2021-08-01                                   sa-on-vip     6695   
37   2021-08-01                                   sa-on-vip     6696   
38   2021-08-01                      search-list-algorithms     5528   
39   2021-08-01                search-pads-none-desktop-mla     3478   
40   2021-08-01                search-pads-none-desktop-mla     3479   
41   2021-08-01                                 secure_card     4612   
42   2021-08-01                    seller-reputation-change     4552   
43   2021-08-01                    seller-reputation-change     4553   
44   2021-08-01                         seller-with-tooltip     4692   
45   2021-08-01                    servicesQuoteUnification  DEFAULT   
46   2021-08-01  shippingCalculatorMigrationModalExperiment     6551   
47   2021-08-01  shippingCalculatorMigrationModalExperiment  DEFAULT   
48   2021-08-01                            show-pads-global     5175   
49   2021-08-01                            show-pads-global     5176   
50   2021-08-01                       show-pads-search-list     5145   
51   2021-08-01                       show-pads-search-list     5146   
52   2021-08-01                         showV2V3BoxMessages     6430   
53   2021-08-01                         showV2V3BoxMessages  DEFAULT   
54   2021-08-01                             sort-by

In [ ]:
df_grouped_00[df_grouped_00.experiment=='checkOnBehavior'].head(2)

date       experiment variant  participants  purchases  buy_rate  \
19  2021-08-01  checkOnBehavior     348            22          0  0.000000   
20  2021-08-01  checkOnBehavior     349            21          1  0.047619   

   buy_rate_percent  
19            0.00%  
20            4.76%

In [ ]:
df_grouped_00.groupby('date')['participants'].max()

date
2021-08-01    1845
2021-08-02    5962
Name: participants, dtype: int64

In [ ]:
df_grouped_00.date = df_grouped_00.date.astype(str)

In [ ]:
df_grouped.to_csv('/Users/juanmanuelpaiba/Documents/Juan_Paiba/AB_Testing_Meli/data/Outputs/data_test.csv', sep = ",", index = False)
df_grouped_00.to_csv('/Users/juanmanuelpaiba/Documents/Juan_Paiba/AB_Testing_Meli/python_scripts/grouped_inf.csv', sep = ",", index = False)
df_grouped_00.to_csv('/Users/juanmanuelpaiba/Documents/Juan_Paiba/AB_Testing_Meli/data/Outputs/grouped_inf.csv', sep = ",", index = False)

In [ ]:
df_grouped.head(9)

date           experiment variant  user_id    path  \
0  2021-08-01  HideTransitionModal    6646   336575  mshops   
1  2021-08-01  HideTransitionModal    6646  3677574  mshops   
2  2021-08-01  HideTransitionModal    6647  2234611  mshops   
3  2021-08-01  HideTransitionModal    6647  3023131  mshops   
4  2021-08-01  HideTransitionModal    6647  9116431  mshops   
5  2021-08-01  HideTransitionModal    6993   161748  mshops   
6  2021-08-01  HideTransitionModal    6993   762044  mshops   
7  2021-08-01  HideTransitionModal    6993  5219912  mshops   
8  2021-08-01  HideTransitionModal    6993  7034551  mshops   

   purchase_funnel_flag  item_id  
0                     0        0  
1                     0        0  
2                     0        0  
3                     0        0  
4                     0        0  
5                     0        0  
6                     0        0  
7                     0        0  
8                     0        0

In [ ]:
#archivo_valido = not df_results.empty

In [ ]:
#archivo_valido

True

In [ ]:

print(df_results.head())

         date           experiment  variant  participants  purchases  buy_rate buy_rate_percent
0  2021-08-01  HideTransitionModal     6646             2          0  0.000000            0.00%
1  2021-08-01  HideTransitionModal     6647             3          0  0.000000            0.00%
2  2021-08-01  HideTransitionModal     6993             5          0  0.000000            0.00%
3  2021-08-01  HideTransitionModal  DEFAULT            12          0  0.000000            0.00%
4  2021-08-01          address_hub     3574           189        143  0.756614           75.66%


In [ ]:
# Filtramos los datos del dataframe para obtener los resultados del día y experimento que deseamos conocer
df_grouped_00[(df_grouped_00['date'] == '2021-08-01') & (df_grouped_00['experiment'] == 'checkOnBehavior')]

date       experiment variant  participants  purchases  buy_rate  \
19  2021-08-01  checkOnBehavior     348            22          0  0.000000   
20  2021-08-01  checkOnBehavior     349            21          1  0.047619   

   buy_rate_percent  
19            0.00%  
20            4.76%

In [ ]:
#pip install flask
df_grouped_00.to_csv('grouped_inf.csv', sep = ",", index = False)

In [ ]:
import os
os.getcwd()

'/Users/juanmanuelpaiba/Documents/Juan_Paiba/AB_Testing_Meli/jupyter_notebooks'

https://learn.microsoft.com/en-us/azure/developer/python/tutorial-containerize-simple-web-app-for-app-service?tabs=web-app-flask

http://pruebamelijmpaiba.azurewebsites.net/experiment/sa-on-vip/result?day=2021-08-01

In [ ]:
df_results = pd.read_csv("grouped_inf.csv", sep=",")

In [ ]:
#app = Flask(__name__)

#@app.route('/experiment/<experiment_id>/result')

#ef get_experiment_results(experiment_id):

    return jsonify({'resultados': f"mensaje + {experiment_id}"})

# Iniicamos la API en el local host
#if __name__ == '__main__':
 #   app.run()

 * Serving Flask app '__main__'
 * Debug mode: off


 * Running on http://127.0.0.1:5000
Press CTRL+C to quit


In [ ]:
#filtered_results = df_results[(df_results['experiment'] == "HideTransitionModal")]

In [ ]:
#filtered_results

date           experiment  variant  participants  purchases  buy_rate buy_rate_percent
0   2021-08-01  HideTransitionModal     6646             2          0  0.000000            0.00%
1   2021-08-01  HideTransitionModal     6647             3          0  0.000000            0.00%
2   2021-08-01  HideTransitionModal     6993             5          0  0.000000            0.00%
3   2021-08-01  HideTransitionModal  DEFAULT            12          0  0.000000            0.00%
71  2021-08-02  HideTransitionModal     6646            17          1  0.058824            5.88%
72  2021-08-02  HideTransitionModal     6647            16          0  0.000000            0.00%
73  2021-08-02  HideTransitionModal     6993            16          1  0.062500            6.25%
74  2021-08-02  HideTransitionModal  DEFAULT            31          0  0.000000            0.00%

In [ ]:
#df_results.head()

date           experiment  variant  participants  purchases  buy_rate buy_rate_percent
0  2021-08-01  HideTransitionModal     6646             2          0  0.000000            0.00%
1  2021-08-01  HideTransitionModal     6647             3          0  0.000000            0.00%
2  2021-08-01  HideTransitionModal     6993             5          0  0.000000            0.00%
3  2021-08-01  HideTransitionModal  DEFAULT            12          0  0.000000            0.00%
4  2021-08-01          address_hub     3574           189        143  0.756614           75.66%